substrate/frame/staking-async/src/pallet/mod.rs

```diff
@@ -538,40 +541,105 @@ pub mod pallet {
+	#[scale_info(skip_type_params(T))]
+	pub struct BoundedExposurePage<T: Config>(pub ExposurePage<T::AccountId, BalanceOf<T>>);
+	impl<T: Config> Deref for BoundedExposurePage<T> {
+		type Target = ExposurePage<T::AccountId, BalanceOf<T>>;
+
+		fn deref(&self) -> &Self::Target {
+			&self.0
+		}
+	}
+
+	impl<T: Config> core::ops::DerefMut for BoundedExposurePage<T> {
+		fn deref_mut(&mut self) -> &mut Self::Target {
+			&mut self.0
+		}
+	}
+
+	impl<T: Config> codec::MaxEncodedLen for BoundedExposurePage<T> {
+		fn max_encoded_len() -> usize {
+			let max_exposure_page_size = T::MaxExposurePageSize::get() as usize;
+			let individual_size =
+				T::AccountId::max_encoded_len() + BalanceOf::<T>::max_encoded_len();
+
+			// 1 balance for `total`
+			BalanceOf::<T>::max_encoded_len() +
+			// individual_size multiplied by page size
+				max_exposure_page_size.saturating_mul(individual_size)
+		}
+	}
+
+	impl<T: Config> From<ExposurePage<T::AccountId, BalanceOf<T>>> for BoundedExposurePage<T> {
+		fn from(value: ExposurePage<T::AccountId, BalanceOf<T>>) -> Self {
+			Self(value)
+		}
+	}
+
+	impl<T: Config> From<BoundedExposurePage<T>> for ExposurePage<T::AccountId, BalanceOf<T>> {
+		fn from(value: BoundedExposurePage<T>) -> Self {
+			value.0
+		}
+	}
+
+	impl<T: Config> codec::EncodeLike<BoundedExposurePage<T>>
+		for ExposurePage<T::AccountId, BalanceOf<T>>
+	{
+	}
+
 	/// Paginated exposure of a validator at given era.
 	///
 	/// This is keyed first by the era index to allow bulk deletion, then stash account and finally
 	/// the page. Should only be accessed through `Eras`.
 	///
 	/// This is cleared after [`Config::HistoryDepth`] eras.
 	#[pallet::storage]
-	#[pallet::unbounded]
 	pub type ErasStakersPaged<T: Config> = StorageNMap<
 		_,
 		(
 			NMapKey<Twox64Concat, EraIndex>,
 			NMapKey<Twox64Concat, T::AccountId>,
 			NMapKey<Twox64Concat, Page>,
 		),
-		ExposurePage<T::AccountId, BalanceOf<T>>,
+		BoundedExposurePage<T>,
 		OptionQuery,
 	>;
 
+	pub struct ErasClaimedRewardsBound<T>(core::marker::PhantomData<T>);
+	impl<T: Config> Get<u32> for ErasClaimedRewardsBound<T> {
+		fn get() -> u32 {
+			let max_total_nominators_per_validator =
+				<T::ElectionProvider as ElectionProvider>::MaxBackersPerWinner::get();
+			let exposure_page_size = T::MaxExposurePageSize::get();
+			max_total_nominators_per_validator
+				.saturating_div(exposure_page_size)
+				.saturating_add(1)
+		}
+	}
+
 	/// History of claimed paged rewards by era and validator.
 	///
 	/// This is keyed by era and validator stash which maps to the set of page indexes which have
 	/// been claimed.
 	///
 	/// It is removed after [`Config::HistoryDepth`] eras.
 	#[pallet::storage]
-	#[pallet::unbounded]
 	pub type ErasClaimedRewards<T: Config> = StorageDoubleMap<
 		_,
 		Twox64Concat,
 		EraIndex,
 		Twox64Concat,
 		T::AccountId,
-		Vec<Page>,
+		WeakBoundedVec<Page, ErasClaimedRewardsBound<T>>,
```

Comment:
Why weak?

Reply:
I am acting based on the assumption that we want to freely tweak `MaxExposurePageSize`, which can change the bound here.